File: app/src/data/knowledge-base.ts
/** `.qaio/knowledge_base/knowledge_base.json` — institutional knowledge entries. */

import schema from "@qaio-ai/agent-schemas/knowledge_base.schema.json";
import { newId, now, readAgentJson, writeAgentJson } from "./agent-file";

export interface KnowledgeEntry {
  id: string;
  title: string;
  content: string;
  source: string;
  tags: string[];
  client?: string;
  project?: string;
  author?: string;
  created_at: string;
  updated_at: string;
}

export interface NewKnowledgeEntry {
  title: string;
  content: string;
  source?: string;
  tags?: string[];
  client?: string;
  project?: string;
  author?: string;
}

export interface KnowledgeEntryUpdate {
  title?: string;
  content?: string;
  source?: string;
  tags?: string[];
  client?: string | null;
  project?: string | null;
  author?: string | null;
}

const NAME = "knowledge_base";
const s = schema as unknown as Parameters<typeof readAgentJson>[2];

export async function list(agentPath: string): Promise<KnowledgeEntry[]> {
  return readAgentJson<KnowledgeEntry[]>(agentPath, NAME, s, []);
}

export async function search(
  agentPath: string,
  query: string,
): Promise<KnowledgeEntry[]> {
  const all = await list(agentPath);
  if (!query) return all;
  const q = query.toLowerCase();
  return all.filter(
    (e) =>
      e.title.toLowerCase().includes(q) ||
      e.content.toLowerCase().includes(q) ||
      e.tags.some((t) => t.toLowerCase().includes(q)) ||
      (e.client ?? "").toLowerCase().includes(q) ||
      (e.project ?? "").toLowerCase().includes(q),
  );
}

export async function create(
  agentPath: string,
  input: NewKnowledgeEntry,
): Promise<KnowledgeEntry> {
  const items = await list(agentPath);
  const ts = now();
  const entry: KnowledgeEntry = {
    id: newId(),
    title: input.title,
    content: input.content,
    source: input.source ?? "manual",
    tags: input.tags ?? [],
    client: input.client,
    project: input.project,
    author: input.author,
    created_at: ts,
    updated_at: ts,
  };
  await writeAgentJson(agentPath, NAME, s, [...items, entry]);
  return entry;
}

export async function update(
  agentPath: string,
  id: string,
  patch: KnowledgeEntryUpdate,
): Promise<KnowledgeEntry> {
  const items = await list(agentPath);
  const idx = items.findIndex((e) => e.id === id);
  if (idx === -1) throw new Error(`Knowledge entry not found: ${id}`);
  const merged: KnowledgeEntry = {
    ...items[idx],
    ...patch,
    client: patch.client === null ? undefined : (patch.client ?? items[idx].client),
    project: patch.project === null ? undefined : (patch.project ?? items[idx].project),
    author: patch.author === null ? undefined : (patch.author ?? items[idx].author),
    updated_at: now(),
  };
  const next = [...items];
  next[idx] = merged;
  await writeAgentJson(agentPath, NAME, s, next);
  return merged;
}

export async function remove(agentPath: string, id: string): Promise<void> {
  const items = await list(agentPath);
  const next = items.filter((e) => e.id !== id);
  if (next.length === items.length)
    throw new Error(`Knowledge entry not found: ${id}`);
  await writeAgentJson(agentPath, NAME, s, next);
}

File: engine/qaio-engine-core/src/agents/knowledge_base.rs
//! CRUD operations for `.qaio/knowledge_base/knowledge_base.json`.

use super::store::{read_json, write_json};
use super::types::{KnowledgeEntry, KnowledgeEntryUpdate, NewKnowledgeEntry};
use crate::error::{CoreError, CoreResult};
use chrono::Utc;
use std::path::Path;
use uuid::Uuid;

const FILE: &str = "knowledge_base";

pub fn list(root: &Path) -> CoreResult<Vec<KnowledgeEntry>> {
    read_json::<Vec<KnowledgeEntry>>(root, FILE)
}

pub fn search(root: &Path, query: &str) -> CoreResult<Vec<KnowledgeEntry>> {
    let all = list(root)?;
    if query.is_empty() {
        return Ok(all);
    }
    let q = query.to_ascii_lowercase();
    Ok(all
        .into_iter()
        .filter(|e| {
            e.title.to_ascii_lowercase().contains(&q)
                || e.content.to_ascii_lowercase().contains(&q)
                || e.tags.iter().any(|t| t.to_ascii_lowercase().contains(&q))
                || e.client.as_deref().unwrap_or("").to_ascii_lowercase().contains(&q)
                || e.project.as_deref().unwrap_or("").to_ascii_lowercase().contains(&q)
        })
        .collect())
}

pub fn create(root: &Path, input: NewKnowledgeEntry) -> CoreResult<KnowledgeEntry> {
    let mut items = list(root)?;
    let now = Utc::now().to_rfc3339();
    let entry = KnowledgeEntry {
        id: Uuid::new_v4().to_string(),
        title: input.title,
        content: input.content,
        source: input.source,
        tags: input.tags,
        client: input.client,
        project: input.project,
        author: input.author,
        created_at: now.clone(),
        updated_at: now,
    };
    items.push(entry.clone());
    write_json(root, FILE, &items)?;
    Ok(entry)
}

pub fn update(root: &Path, id: &str, updates: KnowledgeEntryUpdate) -> CoreResult<KnowledgeEntry> {
    let mut items = list(root)?;
    let entry = items
        .iter_mut()
        .find(|e| e.id == id)
        .ok_or_else(|| CoreError::NotFound(format!("knowledge entry {id}")))?;

    if let Some(title) = updates.title {
        entry.title = title;
    }
    if let Some(content) = updates.content {
        entry.content = content;
    }
    if let Some(source) = updates.source {
        entry.source = source;
    }
    if let Some(tags) = updates.tags {
        entry.tags = tags;
    }
    if let Some(client) = updates.client {
        entry.client = client;
    }
    if let Some(project) = updates.project {
        entry.project = project;
    }
    if let Some(author) = updates.author {
        entry.author = author;
    }
    entry.updated_at = Utc::now().to_rfc3339();
    let result = entry.clone();
    write_json(root, FILE, &items)?;
    Ok(result)
}

pub fn delete(root: &Path, id: &str) -> CoreResult<()> {
    let mut items = list(root)?;
    let len = items.len();
    items.retain(|e| e.id != id);
    if items.len() == len {
        return Err(CoreError::NotFound(format!("knowledge entry {id}")));
    }
    write_json(root, FILE, &items)
}
